compiler/rustc_middle/src/ty/context.rs

```diff
@@ -1719,18 +1752,23 @@ impl<'tcx> TyCtxt<'tcx> {
         self.mk_ty(Array(ty, ty::Const::from_usize(self, n)))
     }
 
+    #[inline]
+    pub fn mk_array_const(self, ty: Ty<'tcx>, ct: Const<'tcx>) -> Ty<'tcx> {
+        self.mk_ty(Array(ty, ct))
+    }
+
     #[inline]
     pub fn mk_slice(self, ty: Ty<'tcx>) -> Ty<'tcx> {
         self.mk_ty(Slice(ty))
     }
 
     #[inline]
     pub fn intern_tup(self, ts: &[Ty<'tcx>]) -> Ty<'tcx> {
-        self.mk_ty(Tuple(self.intern_type_list(&ts)))
+        if ts.is_empty() { self.types.unit } else { self.mk_ty(Tuple(self.intern_type_list(&ts))) }
     }
 
     pub fn mk_tup<I: InternAs<Ty<'tcx>, Ty<'tcx>>>(self, iter: I) -> I::Output {
-        iter.intern_with(|ts| self.mk_ty(Tuple(self.intern_type_list(&ts))))
+        iter.intern_with(|ts| self.intern_tup(ts))
```

Comment:
I guess we have both? Does it make sense to have both?

Reply:
One takes a slice, and one takes an iterator and collects from that into a slice -- it's nice to have the "raw" `intern_` one when you know the input is a slice